{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<center>\n",
    "<h1>Workshop 2: Quality Control (QC)</h1><br><br>\n",
    "<i><big>Preparing and filtering data for genome-wide association studies.</big><br><br>\n",
    "Hatzikotoulas Konstantinos (Kostas) (konstantinos.hatzikotoulas@helmholtz-munich.de)</i>\n",
    "</center>\n",
    "\n",
    "## Objectives\n",
    "\n",
    "In this workshop, you will learn the data quality assessment and control steps that are typically carried out during genome wide association studies (GWAS).\n",
    "\n",
    "## Why do we need the QC?\n",
    "\n",
    "Study design and errors in genotype calling can introduce systematic biases into GWAS, leading to spurious associations.\n",
    "A thorough QC can help us identify samples and markers that should be removed prior to association analysis in order to minimize the number of false-positive and false-negative associations. \n",
    "\n",
    "<img src=\"FN.png\" width=\"40%\">\n",
    "\n",
    "In this tutorial, we assume that the study design has been conducted appropriately and the QC applies to genotypes after they have been called from probe intensity data.\n",
    "\n",
    "## QC protocol\n",
    "\n",
    "The QC protocol of a GWAS is usually split into two broad categories, “Sample QC” and “Variant QC”.\n",
    "Sample QC is done prior to Variant QC because we want to maximise the number of markers remaining in the study.\n",
    "\n",
    "We will be using <code>plink</code> to run the QC and R to visualise the results.\n",
    "You can find a manual and command reference of <code>plink</code> [here](https://www.cog-genomics.org/plink/1.9/) and [here](https://www.cog-genomics.org/plink/1.9/index).\n",
    "\n",
    "\n",
    "## Sample QC\n",
    "\n",
    "It consists of (at least) five steps:\n",
    "\n",
    "<img src=\"QCsteps.png\" width=\"40%\">\n",
    "\n",
    "\n",
    "<br>\n",
    "Initial setup:\n",
    "<br>\n",
    "\n",
    "<br>\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "<b>Location:</b> The data for this workshop can be found in this directory: <code>/home/volos/data/Workshop2_QC</code>. \n",
    "\n",
    "</div>\n",
    "\n",
    "\n",
    "```bash\n",
    "\n",
    "#Make your own directory.\n",
    "mkdir QC_[your name]\n",
    "cd QC_[your name]\n",
    "(eg mkdir QC_Kostas)\n",
    "\n",
    "## test plink runs:\n",
    "plink --help\n",
    "## test that R runs for you, then use q() and then n to exit R\n",
    "R\n",
    "\n",
    "#Give the path of your DIR and initial unqced FILE\n",
    "\n",
    "DIR=<path to your directory>\n",
    "\n",
    "(eg DIR=/home/volos/QC_Kostas)\n",
    "\n",
    "FILE=VSS\n",
    "\n",
    "#Copy the initial UnQCed data to your present working directory (pwd)\n",
    "\n",
    "cd $DIR\n",
    "\n",
    "scp -pr /home/volos/data/Workshop2_QC/Step1/VSS.* ./\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "### Step_1: Individuals with outlying missing genotype\n",
    "\n",
    "#### Call rate\n",
    "\n",
    "##### Run missingness across file genome-wide\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --missing --out $DIR/$FILE-missing\n",
    "\n",
    "```\n",
    "\n",
    "##### Produce a log file giving samples excluded at CR 0.98 to check against R result\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --mind 0.02 --make-bed --out $DIR/$FILE-mind0.02\n",
    "\n",
    "```\n",
    "\n",
    "### Step_2: Individuals with discordant sex information\n",
    "\n",
    "#### Sex check\n",
    "\n",
    "##### Run sex checking\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --check-sex --out $DIR/$FILE-sexcheck\n",
    "\n",
    "```\n",
    "##### Extract xchr SNPs\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --chr 23 --make-bed --out $DIR/$FILE-xchr\n",
    "\n",
    "```\n",
    "\n",
    "##### Run missingness on xchr SNPs\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-xchr --missing --out $DIR/$FILE-xchr-missing\n",
    "\n",
    "```\n",
    "\n",
    "### Step_3: Individuals with outlying heterozygosity rate\n",
    "\n",
    "#### Heterozygosity\n",
    "\n",
    "##### Extract autosomal SNPs\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --autosome --make-bed --out $DIR/$FILE-chr1-22\n",
    "\n",
    "```\n",
    "\n",
    "##### Extract SNPs with minor allele frequency (MAF) greater than/equal to 1%\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22 --maf 0.01 --make-bed --out $DIR/$FILE-chr1-22-mafgte0.01\n",
    "\n",
    "```\n",
    "\n",
    "##### Extract SNPs with MAF less than 1%\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22 --exclude $DIR/$FILE-chr1-22-mafgte0.01.bim --make-bed --out $DIR/$FILE-chr1-22-mafless0.01\n",
    "\n",
    "```\n",
    "##### Get missingness to plot against het in R\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafless0.01 --missing --out $DIR/$FILE-chr1-22-mafless0.01-missing\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafgte0.01 --missing --out $DIR/$FILE-chr1-22-mafgte0.01-missing\n",
    "\n",
    "```\n",
    "\n",
    "##### Convert both to ped/map files for heterozygosity’s script\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafless0.01 --recode --out $DIR/$FILE-chr1-22-mafless0.01-recode\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafgte0.01 --recode --out $DIR/$FILE-chr1-22-mafgte0.01-recode\n",
    "\n",
    "```\n",
    "\n",
    "##### Run heterozygosity’s script\n",
    "\n",
    "```\n",
    "cp /home/volos/data/Workshop2_QC/Step3/calc_het.pl $DIR/\n",
    "perl calc_het.pl -f VSS-chr1-22-mafgte0.01-recode.ped\n",
    "perl calc_het.pl -f VSS-chr1-22-mafless0.01-recode.ped\n",
    "\n",
    "```\n",
    "Need to change the headers in Summary file to get rid of the spaces.\n",
    "\n",
    "ID total num_hom num_het Percent_hom Percent_het\n",
    "\n",
    "\n",
    "### Step_4: Duplicated or related individuals\n",
    "\n",
    "#### Relatedness/Duplicates\n",
    "\n",
    "##### Pair-wise IBD to look at duplicates.Using only variants ≥1%, excluding complex regions and LD prune using R-squared 0.2.\n",
    "\n",
    "##### Exclude complex regions\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafgte0.01 --exclude /home/volos/data/Workshop2_QC/Step4/complex_regions.txt --range --make-bed --out $DIR/$FILE-chr1-22-mafgte0.01-noCR\n",
    "\n",
    "```\n",
    "##### LD prune\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafgte0.01-noCR --indep 50 5 1.25 --out $DIR/$FILE-chr1-22-mafgte0.01-noCR-pruning\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafgte0.01-noCR --extract $DIR/$FILE-chr1-22-mafgte0.01-noCR-pruning.prune.in --make-bed --out $DIR/$FILE-chr1-22-mafgte0.01-noCR-LDpruned0.2\n",
    "\n",
    "```\n",
    "##### Pair-wise IBD\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-chr1-22-mafgte0.01-noCR-LDpruned0.2 --genome --out $DIR/$FILE-chr1-22-mafgte0.01-noCR-LDpruned0.2-genome\n",
    "\n",
    "```\n",
    "##### Run space_to_tab.pl on the genome result\n",
    "\n",
    "```\n",
    "\n",
    "perl /home/volos/data/Workshop2_QC/Step4/space_to_tab.pl VSS-chr1-22-mafgte0.01-noCR-LDpruned0.2-genome.genome\n",
    "\n",
    "```\n",
    "\n",
    "### Step_5: Ethnicity outliers\n",
    "\n",
    "#### Ethnicity MDS distance matrix\n",
    "\n",
    "For this step, you will need to merge your data with 1000 Genomes genotype data (or HapMap genotype data)\n",
    "(Its a time consuming procedure thus please use the following files)\n",
    "\n",
    "```\n",
    "cp /home/volos/data/Workshop2_QC/Step5/VSS-1Kg.bed $DIR/\n",
    "\n",
    "cp /home/volos/data/Workshop2_QC/Step5/VSS-1Kg.bim $DIR/\n",
    "\n",
    "cp /home/volos/data/Workshop2_QC/Step5/VSS-1Kg.fam $DIR/\n",
    "\n",
    "```\n",
    "\n",
    "```\n",
    "\n",
    "FILE2=VSS-1Kg\n",
    "\n",
    "```\n",
    "\n",
    "##### Pair-wise IBD.\n",
    "\n",
    "##### Using only autosomal variants with MAF ≥1%, excluding complex regions.\n",
    "\n",
    "```\n",
    "\n",
    "plink --bfile $DIR/$FILE2 --autosome --maf 0.01 --exclude /home/volos/data/Workshop2_QC/Step4/complex_regions.txt --range --make-bed --out $DIR/$FILE2-chr1-22-mafgte0.01-noCR\n",
    "\n",
    "```\n",
    "\n",
    "##### LD prune using R-squared 0.2.\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE2-chr1-22-mafgte0.01-noCR --indep 50 5 1.25 --out $DIR/$FILE2-chr1-22-mafgte0.01-noCR-pruning\n",
    "plink --bfile $DIR/$FILE2-chr1-22-mafgte0.01-noCR --extract $DIR/$FILE2-chr1-22-mafgte0.01-noCR-pruning.prune.in --make-bed --out $DIR/$FILE2-chr1-22-mafgte0.01-noCR-LDpruned0.2\n",
    "\n",
    "```\n",
    "\n",
    "##### Pair-wise IBD\n",
    "\n",
    "```\n",
    "\n",
    "plink --bfile $DIR/$FILE2-chr1-22-mafgte0.01-noCR-LDpruned0.2 --genome --out $DIR/$FILE2-chr1-22-mafgte0.01-noCR-LDpruned0.2-genome\n",
    "perl /home/volos/data/Workshop2_QC/Step4/space_to_tab.pl VSS-1Kg-chr1-22-mafgte0.01-noCR-LDpruned0.2-genome.genome\n",
    "\n",
    "```\n",
    "\n",
    "##### MDS distance matrix calcualting the first 10 components\n",
    "\n",
    "```\n",
    "\n",
    "plink --bfile $DIR/$FILE2-chr1-22-mafgte0.01-noCR-LDpruned0.2 --read-genome $DIR/$FILE2-chr1-22-mafgte0.01-noCR-LDpruned0.2-genome.genome --cluster --mds-plot 10 --out $DIR/$FILE2-chr1-22-mafgte0.01-noCR-LDpruned0.2-genome-mds\n",
    "\n",
    "```\n",
    "\n",
    "Now, we need to check our results against R plots"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
