{"cells":[{"cell_type":"markdown","metadata":{"id":"O1vAYXGLTzKJ"},"source":["<script>\n","    var code_show=true; //true -> hide code at first\n","\n","    function code_toggle() {\n","        $('div.prompt').hide(); // always hide prompt\n","\n","        if (code_show){\n","            $('div.input').hide();\n","        } else {\n","            $('div.input').show();\n","        }\n","        code_show = !code_show\n","    }\n","    $( document ).ready(code_toggle);\n","</script>"]},{"cell_type":"markdown","metadata":{"id":"W3-Xj12iTzKM"},"source":["<center>\n","<h1>Workshop 1b: Additional UNIX/Bash</h1><br><br>\n","<i><big> File operations, pipes, data reformatting and queries.</big><br><br>\n","Konstantinos Hatzikotoulas</i>\n","</center>\n","\n","**Topics covered in this tutorial**:\n","\n","* Text processing tools: `grep`, `cut`, `tr` and `paste`.\n","* Sed and AWK.\n","* Pipes and redirections.\n","\n","**Topics _not_ covered in this tutorial**:\n","\n","* Variable handling.\n","* Loops.\n","* Conditionals.\n","* Subshells, FIFO pipes and named pipes.\n"]},{"cell_type":"markdown","source":["# Introduction"],"metadata":{"id":"lPmMDUThAia7"}},{"cell_type":"markdown","metadata":{"id":"fIrRwNz8TzKN"},"source":["## Step 1 : Downloading resources from the internet\n","\n","It is easy to access the internet via the command line. For this we use the `wget` tool. We want to  download the complete [GWAS catalog](https://www.ebi.ac.uk/gwas/) maintained by [EBI](https://www.ebi.ac.uk/) that contains all the published genome-wide associations to date. This collection by 20.5.2024 contains contains 6868 publications, 619964 top associations and 86880 full summary statistics. You can perform simple queries on the database using the GWAS catalog website, but you can also download it so you can build you own database.\n"]},{"cell_type":"markdown","source":["### Question 0  \n","Go to the GWAS catalog website and find the download link. Create a directory in your home named `GWAS_catalog` and download it to a file named `GWAS_catalog_2022.11`. How large is the file you just downloaded?\n","\n","\n","Click the arrow key besides the \"Question 0\" header to show/hide the answer."],"metadata":{"id":"K9qQVQRjF8X7"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"LbZ_v7tITzKN"},"outputs":[],"source":["%%bash\n","# We suppress the output messages using the --quiet option\n","# and specify the output file name using the -O option.\n","wget --quiet -O GWAS_catalog.txt https://www.ebi.ac.uk/gwas/api/search/downloads/full\n","\n","# Size\n","ls -lh GWAS_catalog.txt"]},{"cell_type":"markdown","metadata":{"id":"-O3RDTeITzKP"},"source":["`wget` is an extremely powerful bash tool, and has a wide range of amazing features (like resuming interrupted download) that we can't cover here. As it has been mentioned in the previous workshop, you can read the comprehensive manual using `man wget`. This also apply to the rest of the tools we are covering in this session.\n","\n","<div class=\"alert alert-warning\"> The file we just downloaded was created in Windows. This creates problems, because some characters are coded in different ways in UNIX/Linux and Windows. To make the file Linux-compatible, we do `dos2unix GWAS_catalog.txt`.\n","</div>"]},{"cell_type":"markdown","source":["# Text processing bash tools"],"metadata":{"id":"3T1ZI9k5j0VE"}},{"cell_type":"markdown","metadata":{"id":"wVrEzEdcTzKP"},"source":["## Step 2: Counting lines, first $n$ and last $n$ lines of a file"]},{"cell_type":"markdown","source":["### Question 1  \n","Using `wc`, check the number of lines in the GWAS catalog. Using `head` and `tail`, display the first 10 lines and the last 5 lines of the file.\n","\n","Click the arrow key besides the \"Step 2\" header to show/hide the answer.\n"],"metadata":{"id":"vfHubsw5GJbk"}},{"cell_type":"code","execution_count":null,"metadata":{"scrolled":true,"id":"2StY2F8yTzKP"},"outputs":[],"source":["%%bash\n","\n","# How many lines does the file have?\n","wc -l GWAS_catalog.txt\n","\n","# First 10 rows\n","head GWAS_catalog.txt\n","\n","# Last 5 rows\n","tail -5 GWAS_catalog.txt\n","tail -n 5 GWAS_catalog.txt"]},{"cell_type":"markdown","source":["From the first 10 rows we can see how the file is structured:\n","* the first line is the header describing the content of each column (a more comprehensive explanation can be found on the [gwas website](https://www.ebi.ac.uk/gwas/docs/fileheaders))\n","* then each line is an individual association between a genetic variant and an observed phenotype."],"metadata":{"id":"i3vlkImKkgrS"}},{"cell_type":"markdown","metadata":{"id":"gQfdbjr8TzKQ"},"source":["\n","## Step 2b : Using pipes\n","\n","Right now, all the commands we ran were printing their output to the screen (like `wc`) or to a file (`wget -O`). But we can also make the output go to another bash command. We do this using the \"pipe\" character (`|`). For example, `command1 | command2` will pass the output of `command1` to the input of `command2`.\n"]},{"cell_type":"markdown","source":["### Question 2  \n","Using `head` piped to `tail`, print **only the 200th line** of the GWAS catalog.\n","\n","`cat` prints the output of one or more files to the screen. Using `cat` and `wc`, print the number of lines in the GWAS catalog.\n","\n","\n","Click the arrow key besides the \"Question 2\" header to show/hide the answer.\n"],"metadata":{"id":"gZlLzAJtGSS8"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"BBL2z102TzKR"},"outputs":[],"source":["%%bash\n","head -200 GWAS_catalog.txt | tail -1\n","\n","cat GWAS_catalog.txt | wc -l"]},{"cell_type":"markdown","metadata":{"id":"YQHrtl7-TzKR"},"source":["## Step 3 : redirections\n","\n","Instead of outputting the result of a command to another command, we can also tell UNIX to write the output to a file. This is not so useful when you have 1 command, but when you have several piped commands (pipeline) it is used to write the end result to a file. There are 2 types of redirection:\n","\n","* **overwrite** which is a single \"greater than\". `command1 > file` will overwrite `file` if it exists, otherwise create it and fill it with the output of `command1`.\n","* **append**, which is a double \"greater than\". `command1 >> file` will add the output of `command1` to `file`, creating it if it does not exist.\n","\n","\n"]},{"cell_type":"markdown","source":["### Question 3\n","<div class=\"alert alert-success\">  Using `head`, `tail` and pipes, create a file named `200th_line.txt` that contains **only the header and the 200th line**, and display it using `cat`.\n","\n","</div>\n","\n","Click the arrow key besides the \"Step 3\" header to show/hide the answer.\n"],"metadata":{"id":"M7oWPvAEmDa8"}},{"cell_type":"code","execution_count":null,"metadata":{"scrolled":true,"id":"er9IYXB8TzKR"},"outputs":[],"source":["%%bash\n","## First the header\n","head -1 GWAS_catalog.txt > 200th_line.txt\n","\n","## Then the 200th line\n","head -200 GWAS_catalog.txt | tail -1 >> 200th_line.txt\n","\n","cat 200th_line.txt"]},{"cell_type":"markdown","metadata":{"id":"uPWKczvTTzKS"},"source":["### Question 4  \n","The file you have created contains both spaces and tab characters as separators, which makes it a bit messy to display. The `column` command can add artificial spaces to a pipe, so that it looks a bit more like Excel. Pipe `cat` into `column -s$'\\t' -t`, and then into `less`. By default, `less` folds long lines to make them fit the screen. We have very long lines so we want to display them laterally, without breaking them. Can you find the option in the `less` manual that does that? The 200th line of the GWAS catalog tells us about a mutation in a gene. Can you find the chromosome, position and gene for that variant?\n","\n","\n","Click the arrow key besides the \"Question 4\" header to show/hide the answer."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Z84PRD-bTzKS"},"outputs":[],"source":["%%bash\n","cat 200th_line.txt | column -s$'\\t' -t"]},{"cell_type":"markdown","metadata":{"id":"3UroPSHYTzKT"},"source":["### Searching files\n","\n","Very often, we want to search the contents of files. This is done using a very powerful command, `grep`.\n","\n","```bash\n","\n","grep [OPTIONS] [PATTERN] [FILE]\n","\n","# Example:\n","\n","grep Munich cities.txt\n","```\n","\n","the last example looks for every line that contains \"Munich\" in the file `cities.txt` and displays the corresponding row. If your pattern contains a space, enclose it within simple or double quotes (e.g. `\"Munich, Bavaria\"` or `'Munich, Thessalia'`)  \n","\n","It is possible to search for several patterns at a time, not just one. If you are interested in only a few patterns, you can add them to the command line, separating them by `-e`. For example, ` grep -e Munich -e Bavaria cities.txt`. If you have many patterns that you want to search for, you can use `grep -f patterns.txt file.txt` where `patterns.txt` contains all the patterns you are interested in, 1 per line.  \n","\n","Interesting parameters for `grep` are `-w`, `-v`, `-i`, `-n`, `-A`, `-B`, `-C`, `-l` and `-L`. Check them out and play around with them if you have time!\n"]},{"cell_type":"markdown","source":["### Question 5\n","\n","<div class=\"alert alert-success\">Use `grep` to search for other variants in the gene you found above. Using a pipe and `wc`, determine how many lines from the GWAS catalog mention it. How about the variant rs11205303? Are there other lines that mention this variant?\n","</div>"],"metadata":{"id":"XF_tnuvhqLIJ"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"vKDNOHkpTzKT"},"outputs":[],"source":["%%bash\n","## Question 5\n","grep MTMR11 GWAS_catalog.txt | wc -l\n","\n","grep rs11205303 GWAS_catalog.txt | wc -l"]},{"cell_type":"markdown","source":["### Question 6\n","Before, we used 2 steps to create a file containing the 200th line. Now with `grep`, we can do everything in one line. Find something unique to the header and our 200th line, and create a file called `200th_line2.txt` that contains them (you might need several grep commands piped together). Use the `diff` command to check that the files are identical."],"metadata":{"id":"rVq_nJBWp_kT"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"RjmvvitBTzKT"},"outputs":[],"source":["%%bash\n","## Question 6\n","\n","# the -n option of grep is useful here to show which line has a match\n","\n","# There is no single right answer here, you might have selected other search criteria that give the same result.\n","rm 200th_line2.txt\n","fgrep -e DOWNSTREAM_GENE_DISTANCE -e 6.154901959985743 GWAS_catalog.txt > 200th_line2.txt\n","ls\n","head 200th_line2.txt\n","head 200th_line.txt\n","# The diff output is empty: our files are identical!\n","diff 200th_line.txt 200th_line2.txt >"]},{"cell_type":"markdown","metadata":{"id":"9gyX5vdkTzKT"},"source":["## Step 4: Playing with columns: cut and paste\n","\n","As we now know, our file contains many columns, but we are interested in only a couple of them. `cut` allows to extract specific columns from a file, whereas `paste` appends several columns together. The delimiter (space, comma, tab) that separates your fields is specified with the `-d` argument (as they are special characters they will need to be quoted with `'`).\n","\n","#### Example\n","The below command extracts columns 1 to 4, 6, and 8 to 10 of `file.txt` and prints the output to the screen.\n","```bash\n","cut -f1-4,6,8-10 file.txt\n","```"]},{"cell_type":"markdown","source":["### Question 7  \n","Extract the associated disease/trait and the SNP id (column `SNPS`) from the GWAS catalog and write them to a file called `phenotype_SNP.txt`. Then create a file called `SNP_phenotype.txt` where the SNP column comes first, and the Disease/trait one second (`cut` alone is not enough for this, you need to create 2 files and paste them together)."],"metadata":{"id":"44NERfRs7dJw"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"-aQNhDMyTzKU"},"outputs":[],"source":["%%bash\n","cut -f8,22 GWAS_catalog.txt > phenotype_SNP.txt\n","\n","cut -f8 GWAS_catalog.txt > phenotype.txt\n","cut -f22 GWAS_catalog.txt > SNP.txt\n","\n","paste SNP.txt phenotype.txt > SNP_phenotype.txt\n","\n","echo -- Content of phenotype_SNP.txt --\n","head -n 4 phenotype_SNP.txt\n","echo\n","echo -- Content of phenotype_SNP.txt --\n","head -n 4 SNP_phenotype.txt"]},{"cell_type":"markdown","metadata":{"id":"99-bmYWBTzKU"},"source":["## Step 5 : Find and replace\n","\n","Once we have extracted some data, we often want to modify it. There are 2 programs to do this type of thing, one very simple and the other very complicated:\n","\n","* `tr` replaces every occurrence of one particular character by another character\n","  * The syntax is very simple : `command1 | tr 'a' 'b' ` replaces every occurrence of the letter a in the pipe by the letter b.\n","* `sed` is more complicated to use, but allows to substitute pretty much anything by pretty much anything else.\n","  * `command1 | sed 's/pattern_to_find/replacement pattern/'`\n","  * `pattern_to_find` is a **regular expression**, a special command that allow you to match certain bits of text and not others. Regular expressions are very, very common in the computing world, and many programs understand them (`grep` also allows them). Regular expressions (or regexes) can take a long time to learn and use properly. You can learn about them [here](http://www.regular-expressions.info/) and test them out [here](http://regexr.com/).\n","\n","### Some (very) basic regular expressions\n","\n","* `/Munich/` matches the string `Munich`, as expected\n","* `/Mun*ich/` matches Munich with zero or more `n`: `Munich`, `Muich` `Munnnnnnich`\n","* `/Mun+ich/` matches Munich with one or more `n`: `Munich`, `Munnnnnich` but not `Muich`\n","* `/Mu.ich/` matches Munich, but the third letter can be anything (but not nothing): `Munich`, `Muyich`, `Muvich` but not `Munnich`\n","* `/$/` matches the end of the line\n","* `/^/` matches the beginning of the line\n","* `/^Munich$/` matches a line that contains only `Munich` but not if there is a space (or anything) before or after.\n","* `s/n/i/` will replace the first occurrence of `n` to `i`, so `Munich` to `Muiich`.\n","* By adding a \"global\" flag like `s/a/i/g` (the `g` is the global flag), this changes the behaviour to replace all occurrences instead of one. So `s/a/i/g` will change `Bavaria` to `Bivirii`.\n","* etc..."]},{"cell_type":"markdown","source":["### Question 8 & 9\n","\n","<b>Question 8:</b> Using `tr`, replace every space character in the GWAS catalog by an underscore (`_`) and write it to `GWAS_catalog_no_spaces.txt`. Remember they are special characters, so need to be quoted.\n","</div>\n","\n","<b>Question 9:</b> In the file `phenotype_SNP.txt` you generated before, some SNPs have several IDs separated by a semicolon. Find them and remove all alternate IDs, keeping only the first one; write the file to `duplicates_removed.txt`.\n","</div>\n"],"metadata":{"id":"YXXu71v6tv93"}},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"id":"VUs5dzpBTzKU"},"outputs":[],"source":["%%bash\n","## Question 8\n","head GWAS_catalog.txt\n","cat GWAS_catalog.txt | tr ' ' '_' > GWAS_catalog_no_spaces.txt\n","head GWAS_catalog_no_spaces.txt\n","\n","## Question 9\n","# It's not enough to just search for ';', because there are trait column\n","# values with semi-colons included (e.g. 'Serum levels of protein TPSB2;TPSAB1').\n","# So we write a simple regex to search\n","grep -e '[0-9]; ' phenotype_SNP.txt | sed 's/;.*//' > duplicates_removed.txt\n","\n","head -n 5 duplicates_removed.txt"]},{"cell_type":"markdown","metadata":{"id":"14mywuQATzKV"},"source":["## Step 6 : Advanced column manipulation using `awk`\n","\n","Here, we introduce another very powerful tool: `awk`. Like `sed`, entire programs can be written in awk, so we will show only a tiny fraction of what can be done with it.\n","\n","The great usefulness of `awk` is that it **automatically splits lines on whitespace**, allowing you to select specific columns and performing actions on them.\n","\n","<b>Beware:</b> `awk` splits on **any whitespace**. What will happen in a file like ours, with both tabs and spaces? Try it out: run `awk '{print $2}' GWAS_catalog.txt | less`. Look at the first row. Is this what you would expect? What would you do to solve this problem?\n","\n","### Some (very) basic examples\n","\n","\n","* `'{print $2}'` prints the second field\n","* `'{print $0}'` prints the whole line\n","* `'$2==1'` prints the whole line if the second field is equal to 1 (equivalent to `'{if($2==1){print $0}}'`)\n","* `'$3~/Munich/{print \"field 2 is: \", $2+1, \"field 3 is: \", $3}'` prints a custom string if the third field matches (contains) \"Munich\"\n","* `'NF>2'` prints the line if it has more than 2 fields\n","* `'$NF>2'` prints the line if the value of the last field is greater than 2\n","* `'$10>$2+1 && $(NF-1)==\"yes\"'` prints the line if the 10th field is greater than the second + 1 **and** if the before-last field is equal to `yes`.\n","\n","\n","<br><br>\n","AWK commands, like `sed` ones, are enclosed within quotes (`'`). If you want to print the whole line only if a condition is satisfied, just write the condition. If you want to print certain columns, or do more complex operations, you must include your code in curly brackets (`{...}`).\n","\n","\n","<br><br>\n","AWK merges consecutive delimiters into one. For example, in the string `a\\tb\\t\\tc`, `$2==\"b\"` and `$3==\"c\"`. Sometimes (as here) we might have empty fields, so we want to keep consecutive delimiters separated. To do that you have to tell awk specifically : `awk -F'\\t' '...'`\n"]},{"cell_type":"markdown","source":["### Question 10\n","Use the version of the GWAS catalog with no spaces we just created. How many fields does the GWAS catalog contain? Using `paste`, `tr` and `seq`, build a file named `column_indices.txt` that contains the column names and their numbers. With `grep`, locate the number of the columns `DISEASE/TRAIT`, `CHR_ID`, `CHR_POS`, `MAPPED_GENE`, `SNPS` in that file. Use awk to extract all records of the GWAS catalog that are on chromosome 11 between 5220000 and 5300000, and print only the fields above to a file called `Hemoglobin_region.txt`.\n"],"metadata":{"id":"Jead2_kwv0bc"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"bzz3lJwgTzKV"},"outputs":[],"source":["%%bash\n","# number of fields (=34)\n","echo \"Number of fields\"\n","head -1 GWAS_catalog.txt| tr ' ' '_' | awk '{print NF}'\n","echo\n","\n","# building a file with column numbers\n","head -1 GWAS_catalog_no_spaces.txt | tr '\\t' '\\n' > header.column\n","seq 1 34 > indices.column\n","paste header.column indices.column > header.columns\n","\n","# finding column IDs\n","echo \"Column numbers\"\n","grep -w -e DISEASE -e CHR_ID -e CHR_POS -e MAPPED_GENE -e SNPS header.columns\n","\n","# Printing only selected fields\n","awk -F'\\t' '$12==11 && $13>5220000 && $13<5300000{print $8, $12, $13, $15, $22}' GWAS_catalog_no_spaces.txt > Hemoglobin_region.txt"]},{"cell_type":"markdown","metadata":{"id":"lrYii0gMTzKV"},"source":["## Step 7:  Sorting\n","\n","Last but not least, we introduce the `sort` and `uniq` commands. Very often we need to sort files according to one or several columns, this is achieved using sort. Sort understands alphabetical, numerical and \"natural\" (i.e. scientific) orders and can compute unique values.\n","\n","#### Examples\n","* `sort -k1,1 -k2,2n` sorts according to the first field (alphanumeric order) then the second field (numeric order)\n","* `sort` sorts on the whole line according to alphanumeric order\n","* `sort -r` sorts randomly\n","\n","<br>\n","`uniq` does not add much to sort, as `sort -u` is the same as `sort | uniq`. `uniq` is frequently used for its `-c` argument, which counts the number of occurrences of every unique line.\n"]},{"cell_type":"markdown","source":["### Question 11\n","How many unique genes are there in our previous query? How many times is each of them mentioned?"],"metadata":{"id":"SCXwLD19wksg"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"jxxYufbzTzKV"},"outputs":[],"source":["%%bash\n","cut -d' ' -f4 Hemoglobin_region.txt | sort | uniq -c"]},{"cell_type":"markdown","metadata":{"id":"nT7QKoKxTzKV"},"source":["## Step 8 : Variables and loops\n","\n","In bash, we can define variables to store temporary information:\n","\n","```bash\n","city=Munich\n","```\n","\n","To recall the value of the variable we have to prefix it with a dollar sign (`$`):\n","```bash\n","echo $city\n","```\n","\n","### Variable operations\n","\n","#### Concatenation\n","This is the easiest of all operations: in order to \"glue\" the value of 2 variables together, simply write them one after the other:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"6hwPhfmwTzKW"},"outputs":[],"source":["%%bash\n","a=1\n","b=2\n","echo $a$b"]},{"cell_type":"markdown","source":["You can also mix variables with text of your own:"],"metadata":{"id":"Sxk269Wm_Qgp"}},{"cell_type":"code","source":["%%bash\n","city=Munich\n","\n","echo Konstantinos lives in $city."],"metadata":{"id":"NN5Up5GH_RxG"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zAve4wMvTzKW"},"source":["<b>Beware:</b> The character that follows your variable name is important. Bash will think that some characters are part of the variable name, for example you cannot write `$a-$b`, although you can write `$a.$b`. If you encounter such characters, you should write `${a}-${b}`.\n","\n","\n"]},{"cell_type":"markdown","source":["#### Mathematical operations\n","\n","* Addition `$(( a + b ))`, `$(( 1 + 2 ))`\n","* Other arithmetic operators `$(( a * b ))`, `$(( a / 10 ))`"],"metadata":{"id":"nYErYEJp_dES"}},{"cell_type":"code","source":["%%bash\n","\n","a=1\n","b=2\n","echo Addition: $(( a + b ))\n","echo Multiplication: $(( a * b ))\n","echo Division: $(( a / b ))"],"metadata":{"id":"HKo7k-ae_fq4"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#### For loops\n","\n","For loops are used to iterate on a list of values:"],"metadata":{"id":"sui2Scck_aoQ"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"fV0tU4ijTzKW"},"outputs":[],"source":["%%bash\n","for i in a b c; do\n","    echo $i\n","done"]},{"cell_type":"markdown","metadata":{"id":"g4F6BA-6TzKW"},"source":["The value of a variable can contain spaces, or other separators. Bash interprets this as a list, so you can iterate over the values in a variable. For example:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PMa6IHp6TzKX"},"outputs":[],"source":["%%bash\n","values=\"a b c d e f g h\"\n","for v in $values; do\n","    echo $v\n","done"]},{"cell_type":"markdown","metadata":{"id":"SaDEWgh9TzKX"},"source":["#### Special variables\n","\n","There are many \"special\" variables hidden in BASH. Some of them are set by the system, such as `$PATH`, which contains all the directories in which your system is going to look for commands. (a funny way to temporarily break your system is to set `PATH` to something random).\n","\n","Another special variable is the range operator:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9Db_eZ-sTzKX"},"outputs":[],"source":["%%bash\n","echo {1..22}"]},{"cell_type":"markdown","metadata":{"id":"GHoY56zATzKX"},"source":["#### The `read` command\n","\n","The `read` command is a very powerful tool which reads whatever it receives through a pipe into a variable. We use it a lot in combination with the `while` loop when we want to **execute the same command on every line of a file**."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"noPQmOVuTzKX"},"outputs":[],"source":["%%bash\n","echo -e \"a\\nb\\nc\\nd\\ne\\nf\\ng\" > test.txt\n","cat test.txt"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"c1WUXCzyTzKX"},"outputs":[],"source":["%%bash\n","cat test.txt | while read letter; do\n","    echo $letter\n","done"]},{"cell_type":"markdown","metadata":{"id":"D4qmiL4zTzKY"},"source":["`read` can also read several fields at the same time:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"pZRSQJ0iTzKY"},"outputs":[],"source":["%%bash\n","paste test.txt test.txt"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Rl3zOlB-TzKY"},"outputs":[],"source":["%%bash\n","paste test.txt test.txt | while read first second; do\n","    echo $first $second $second $first\n","done"]},{"cell_type":"markdown","metadata":{"id":"H8mhSL-lTzKY"},"source":["The great use of this is that you can put whatever you want between `do` and `done` (many many commands, potentially). If you just have one quick command that you want to run, you can use `xargs` instead."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hFbKg-QeTzKY"},"outputs":[],"source":["%%bash\n","cat test.txt | xargs echo"]},{"cell_type":"markdown","metadata":{"id":"rQ1qw9kKTzKY"},"source":["### Question 12 & 13\n","\n","<b>Question 12:</b>  \n","Extract first 10 reported chromosome and position columns from the `GWAS_catalog.txt` file.\n","\n","\n","<b>Question 13:</b>  \n","For every chromosome, position (e.g. `1 1234`), convert this to a list of intervals extended by 1000 on either side (e.g. `1:234-2234`)."]},{"cell_type":"code","source":["%%bash\n","\n","head -n 11 GWAS_catalog_no_spaces.txt | tail -n 10 | cut -f12,13 | while read chrom pos; do\n","  echo \"${chrom}:$(( pos - 1000 ))-$(( pos + 1000 ))\"\n","done"],"metadata":{"id":"BPVF1bgM9TBI"},"execution_count":null,"outputs":[]}],"metadata":{"hide_input":false,"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.4.3"},"colab":{"provenance":[],"collapsed_sections":["3UroPSHYTzKT"]}},"nbformat":4,"nbformat_minor":0}